File: clients/java/Tile.java
import com.sun.jna.Pointer;

///Represents a single tile on the map, can contain some amount of water.
class Tile extends Mappable
{
  public Tile(Pointer p)
  {
    super(p);
  }
  boolean validify()
  {
    if(iteration == BaseAI.iteration) return true;
    for(int i = 0; i < BaseAI.tiles.length; i++)
    {
      if(BaseAI.tiles[i].ID == ID)
      {
        ptr = BaseAI.tiles[i].ptr;
        iteration = BaseAI.iteration;
        return true;
      }
    }
    throw new ExistentialError();
  }

    //commands

  ///Attempt to spawn a unit of a type on this tile.
  boolean spawn(int type)
  {
    validify();
    return (Client.INSTANCE.tileSpawn(ptr, type) == 0) ? false : true;
  }

    //getters

  ///Unique Identifier
  public int getId()
  {
    validify();
    return Client.INSTANCE.tileGetId(ptr);
  }
  ///X position of the object
  public int getX()
  {
    validify();
    return Client.INSTANCE.tileGetX(ptr);
  }
  ///Y position of the object
  public int getY()
  {
    validify();
    return Client.INSTANCE.tileGetY(ptr);
  }
  ///The owner of the tile.
  public int getOwner()
  {
    validify();
    return Client.INSTANCE.tileGetOwner(ptr);
  }
  ///The type of tile this tile represents.
  public int getType()
  {
    validify();
    return Client.INSTANCE.tileGetType(ptr);
  }
  ///Determines if this tile is a part of a Pump Station.
  public int getPumpID()
  {
    validify();
    return Client.INSTANCE.tileGetPumpID(ptr);
  }
  ///The amount of water contained on the tile.
  public int getWaterAmount()
  {
    validify();
    return Client.INSTANCE.tileGetWaterAmount(ptr);
  }
  ///Whether the tile is a trench or not.
  public int getIsTrench()
  {
    validify();
    return Client.INSTANCE.tileGetIsTrench(ptr);
  }

}

File: clients/csharp/Tile.cs
using System;
using System.Runtime.InteropServices;


///Represents a single tile on the map, can contain some amount of water.
public class Tile: Mappable
{

  public Tile()
  {
  }

  public Tile(IntPtr p)
  {
    ptr = p;
    ID = Client.tileGetId(ptr);
    iteration = BaseAI.iteration;
  }

  public override bool validify()
  {
    if(iteration == BaseAI.iteration) return true;
    for(int i = 0; i < BaseAI.tiles.Length; i++)
    {
      if(BaseAI.tiles[i].ID == ID)
      {
        ptr = BaseAI.tiles[i].ptr;
        iteration = BaseAI.iteration;
        return true;
      }
    }
    throw new ExistentialError();
  }

    //commands

  ///Attempt to spawn a unit of a type on this tile.
  public bool spawn(int type)
  {
    validify();
    return (Client.tileSpawn(ptr, type) == 0) ? false : true;
  }

    //getters


  ///Unique Identifier
  public new int Id
  {
    get
    {
      validify();
      int value = Client.tileGetId(ptr);
      return value;
    }
  }

  ///X position of the object
  public new int X
  {
    get
    {
      validify();
      int value = Client.tileGetX(ptr);
      return value;
    }
  }

  ///Y position of the object
  public new int Y
  {
    get
    {
      validify();
      int value = Client.tileGetY(ptr);
      return value;
    }
  }

  ///The owner of the tile.
  public int Owner
  {
    get
    {
      validify();
      int value = Client.tileGetOwner(ptr);
      return value;
    }
  }

  ///The type of tile this tile represents.
  public int Type
  {
    get
    {
      validify();
      int value = Client.tileGetType(ptr);
      return value;
    }
  }

  ///Determines if this tile is a part of a Pump Station.
  public int PumpID
  {
    get
    {
      validify();
      int value = Client.tileGetPumpID(ptr);
      return value;
    }
  }

  ///The amount of water contained on the tile.
  public int WaterAmount
  {
    get
    {
      validify();
      int value = Client.tileGetWaterAmount(ptr);
      return value;
    }
  }

  ///Whether the tile is a trench or not.
  public int IsTrench
  {
    get
    {
      validify();
      int value = Client.tileGetIsTrench(ptr);
      return value;
    }
  }

}
